qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/componentinstaller/components/RestAreaGuideMap.kt

```diff
@@ -0,0 +1,36 @@
+package com.mapbox.navigation.qa_test_app.view.componentinstaller.components
+
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.internal.extensions.flowRouteProgress
+import com.mapbox.navigation.core.internal.extensions.flowRoutesUpdated
+import com.mapbox.navigation.ui.base.lifecycle.UIComponent
+import com.mapbox.navigation.ui.maps.guidance.restarea.api.MapboxRestAreaApi
+import com.mapbox.navigation.ui.maps.guidance.restarea.view.MapboxRestAreaGuideMapView
+
+@ExperimentalPreviewMapboxNavigationAPI
+class RestAreaGuideMap(
+    private val restAreaApi: MapboxRestAreaApi,
+    private val restAreaView: MapboxRestAreaGuideMapView,
+) : UIComponent() {
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        mapboxNavigation.flowRouteProgress().observe { routeProgress ->
+            restAreaApi.generateUpcomingRestAreaGuideMap(routeProgress) {
+                restAreaView.render(it)
+            }
+        }
```

Comment:
`RestStop` can be embedded either in `route.leg.step.bannerInstructions` or `route.leg.step.intersections()`. Hence we have 2 APIs supported in `MapboxRestAreaApi` as follows 

using bannerInstructions
```
fun generateRestAreaGuideMap(
    instructions: BannerInstructions,
    consumer: MapboxNavigationConsumer<Expected<RestAreaGuideMapError, 
    RestAreaGuideMapValue>>
)
```

using RouteProgress
```
fun generateUpcomingRestAreaGuideMap(
    routeProgress: RouteProgress,
    consumer: MapboxNavigationConsumer<Expected<RestAreaGuideMapError, 
    RestAreaGuideMapValue>>
)
```

This is a nice component to have, but based on the logic it only renders `rest_stop.guideMap` if the data is available via `generateUpcomingRestAreaGuideMap(routeProgress)`. We should modify the logic to make it so that it can render rest stops when it is embedded in banner instructions as well. 

The logic can be changed as following:

```
mapboxNavigation.flowRouteProgress().observe { routeProgress ->
    val progressJob = async {
        routeProgress.let { restAreaApi.generateUpcomingRestAreaGuideMap(it) }
    }
    val progressJob = async {
        routeProgress.bannerInstructions?.let { restAreaApi.generateRestAreaGuideMap(it) }
    }
    val progressResult = progressJob.await()
    val result = if (progressResult.isValue) {
        instructionsJob.cancel()
        progressResult
    } else {
        instructionsJob.await()
    }
    result?.let {
        restAreaView.render(it)
    }
}

private suspend fun MapboxRestAreaApi.generateUpcomingRestAreaGuideMap(
    progress: RouteProgress,
): Expected<RestAreaGuideMapError, RestAreaGuideMapValue> {
    return suspendCancellableCoroutine { generateUpcomingRestAreaGuideMap(progress, it::resume) }
}

private suspend fun MapboxRestAreaApi.generateRestAreaGuideMap(
    instructions: BannerInstructions,
): Expected<RestAreaGuideMapError, RestAreaGuideMapValue> {
    return suspendCancellableCoroutine { generateRestAreaGuideMap(instructions, it::resume) }
}
```
Once we update the logic as above, we can use the same activity to test routes that
- contains rest stop in bannerInstructions only
- contains rest stop in step.intersections only
- contains rest stop in both bannerInstructions and step.intersections

We could probably change the spinner to select routes containing rest stops based on bannerInstructions or step.intersections or both instead of short/long SAPA route.

Reply:
Won't do. To be picked as follow-up.